### DISCUSSION TITLE
Debugging and Foreach loops

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#655] [Ideas]
Can we get some minor enhancements to ForEach iterators? ...mostly for debugging purposes.

When debugging some code in VBA that's doing processing loops, I find myself occasionally sincerely wishing (during Edit & Continue debugging) to be able to navigate forward & backwards along the foreach iterator chain sometimes.
i.e. "Oh that data issue was in the previous item in the collection(/whatever), not this current value of N..." if I could only Set N = PreviousForEach(x) in this debugging session...

...but alas, I know of no way to do that, as the iterator variable does not have any accessible member functions with .PreviousInstance or .NextInstance -type functionality to call upon.

Think we could craft a way to accomplish that? (again, for debugging purposes only)

**Comment by WaynePhillipsEA** [comment]
I'm interested;  is this a feature you've come across in others debuggers?

It would be difficult to achieve because the IEnumVARIANT interface (which is used extensively by the For-Each feature) only goes in a forward direction, via the Next method (https://docs.microsoft.com/en-us/windows/win32/api/oaidl/nn-oaidl-ienumvariant).

**Reply by mburns08109** [reply]
Just an idle question here: is there some reason why iEnumVARIANT MUST be the interface we use for iterators under the FOR EACH hood?
Couldn't we substitutes a super class of that interface in its place - one where directional navigation(s) across the underlying selection is permitted (though the default behavior matches iEnumVARIANT's existing actions for FOR EACH) ?

**Reply by bclothier** [reply]
Please first explain to me how you will make a generic spoke that can take a 64" tire that can fit all kinds of bicycles, tricycles, unicycles, and even mopeds or motorcycles of all kind of sizes, with fenders or without fenders and without causing riding issues and I'll explain to you how we can implement such class to enumerate any kind of objects.

**Reply by mburns08109** [reply]
Well that's simple...Nanobots form the spoke and wheel, adjusting their packing algorithms to adjust the structure of the wheel to fit each application required.

ask a stupid question...win stupid answers as prizes, Ben (you did not specify that currently available commercial technology was a restriction).

OK, your turn.  :-)

**Reply by bclothier** [reply]
Having untethered myself from this mundane physical reality, I shall construct an infinite improbability enumerator powered by quantum fluctuations with my sonic  screwdriver and pesto. After, I shall fly the yonder blue sky unfettered. 

Whee! Imaginary solutions is so much more fun than real solutions!

**Reply by mansellan** [reply]
Semi-related: [92: Custom Enumerators](https://github.com/WaynePhillipsEA/twinbasic/issues/92) does discuss exposing a friendlier interface for this (not least because IEnumVARIANT is not implementable in VBx \ TB without tremendous hackery).

But it would still (need to) be forward-only.

**Comment by mburns08109** [comment]
NO - not seen it anywhere to date.
Ach...I was hoping that the under-the-hood API had both forward and backwards (hidden?) methods, so it might've been fairly easy to make them available.

Ah well.  If you feel the need to play with the thought, don't let me stop you, but...perhaps a version 3 thing?  ;-)

**Reply by mansellan** [reply]
`For Each` is a compiler convenience over the [Iterator Pattern](https://en.wikipedia.org/wiki/Iterator_pattern), which is strictly forward-only. In theory, the debugger could cache previous values, but that would then change the memory usage profile of debug vs release code, so... not sure. Maybe a switch?

**Reply by WaynePhillipsEA** [reply]
The debugger could potentially call Reset() on the IEnumVARIANT, then enumerate the collection again to determine the index position of the current value, and then Reset() the collection once again and Skip() to the (index-1) position which would put the IEnumVARIANT then in the correct position...

... nah, too much hassle, plus I'm not convinced it would be worth the effort.

**Reply by bclothier** [reply]
Agreed - even if it was implemented, it would introduce new problems. For example:

```
Dim s As String
For Each foo In bar
  s = s & foo
Next
```

If we go back one iteration, the `s` is still going to contain the values that was added in the current iteration and thus create something that looks like a bug but is actually a mistake on the developer's part. Best to encourage the developer to reset and test with a clean slate.

**Comment by Kr00l** [comment]
It's not mandatory for to handle IEnumVARIANT::Reset.
I have implementations where i just return there E_NOTIMPL and it's working as long as IEnumVARIANT::Next is functional.

**Reply by WaynePhillipsEA** [reply]
Yes, that's true as well.  TBH, I don't think we use Reset/Skip/Clone at all in tB (likely same in VBx), so I bet most implementations don't bother with them.

**Reply by mburns08109** [reply]
OK, so making a multidirectional enumeration is a no-go.
How about a way to expose the enumeration sequence and current index in the sequence? (say, as one of those new associatve arrays?)

Yeah, I know...same sort of issues. but If'n ya don't ask...

**Reply by mansellan** [reply]
I think it's fair to say this is not likely to happen. Iterators just don't work like that.

**Reply by mansellan** [reply]
> Yes, that's true as well. TBH, I don't think we use Reset/Skip/Clone at all in tB (likely same in VBx), so I bet most implementations don't bother with them.

@WaynePhillipsEA I bet that's why .Net simplified the abstraction - nearly nothing uses those methods.

That said, I do wonder what happens when you debug a loop and drag execution back to before the start. Does it reset, or does it clear the internal reference and start again?

**Reply by mansellan** [reply]
I can see what the original thinking was - if you have a sequence where the item length is fixed, you could reset, skip ahead or clone the current position very cheaply. But I suspect that these days, the opportunities for that are very limited.

**Reply by mansellan** [reply]
Hmm... actually...

Is this different between managed and native models? .Net doesn't generally guarantee that objects will stay at the same virtual address over time. Unless you pin the reference, the GC is free to relocate objects during collection so as to defragment the heap.

IIUC, reference-counted languages don't do this, so maybe Skip and Clone make more sense in native environments? I could be talking rubbish of course...
